### PR TITLE
[Tune; RLlib] Missing stopping criterion should not error (just warn).

### PR DESCRIPTION
**Pull request by sven1977**
Missing stopping criterion should not error (just warn).

Some RLlib algorithms take a few iterations in order to sample enough data from their EnvRunners to complete an RL episode. Only after that point in time, episode stats will be published in the result dicts returned by `Algorithm.train()` (`Algorithm` is-a `tune.Trainable`).

The current hacky fix is to prophylactically add common episode stats (such as `episode_return_mean`) to the initialized algorithm's result dict as NaNs, however, this is not a sustailable solution as users might add their own stats and will then run into the `TuneError(Stopping criteria ... not provided)` error without any idea how to fix this. A one-time warning here would be the better solution.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by sven1977** [comment]
Hey @justinvyu , could you take a look at this PR?

**Review by justinvyu — COMMENTED** [review]
Thanks! I think this generally makes sense.

Could we just add a small unit test here, asserting that the warning is logged?

https://github.com/ray-project/ray/blob/master/python/ray/tune/tests/test_trial.py#L5

File: python/ray/tune/experiment/trial.py [hunk 1]
```diff
@@ -851,18 +852,19 @@ def should_stop(self, result):
...
+            if isinstance(criterion, dict):
                 raise ValueError(
                     "Stopping criteria is now flattened by default. "
                     "Use forward slashes to nest values `key1/key2/key3`."
                 )
-            elif result[criteria] >= stop_value:
+            elif criterion not in result:
+                if log_once("tune_trial_stop_criterion_not_found"):
+                    logger.warning(
+                        f"Stopping criterion {criterion} not found in result dict. "
```
[the remark this answers is not in the document]
Comment (sven1977) [reply python/ray/tune/experiment/trial.py]: done

File: python/ray/tune/tests/test_trial.py [hunk 2]
```diff
@@ -116,5 +117,52 @@ def test_trial_logdir_length():
...
+        stub=True,
+        trial_id="abcd1234",
+        stopping_criterion={"a": 10.0, "b/c": 20.0},
+    )
+
+    # Criterion is not reached yet -> don't stop.
+    result = _TrainingResult(
+        checkpoint=None, metrics={"a": 9.999, "b/c": 0.0, "some_other_key": True}
+    )
+    assert not trial.should_stop(result.metrics)
```
Comment (justinvyu) [inline python/ray/tune/tests/test_trial.py:133]: Can we just pass in a dict without the `_TrainingResult` wrapper?

**Review by justinvyu — COMMENTED** [review]
Thanks! Seems like the caplog is not working -- you may need to include the `propagate_logs` fixture.

File: python/ray/tune/tests/test_trial.py (hunk 2 above)
Comment (sven1977) [reply python/ray/tune/tests/test_trial.py:133]: done

**Review by justinvyu — APPROVED** [review]
Thanks!

**Pull request merged**